### ISSUE TITLE
performance problems cause safari crash?

### ISSUE THREAD

**Issue by maggiehe** [vuejs/vue#4365]
### Vue.js version
2.0.2

### Reproduction Link

(performance issue)

### Steps to reproduce

In our root vue instance, the element to mount on sometimes has 250+ child nodes.  (because we use server rendering template)

```
new Vue({
  el: '#body'
})
```

### What is Expected?

Pages can work normally.

### What is actually happening?

The compile process is very slow.

on my phone (iphone6s, ios10.1.1):

- compile 100ms in safari

another phone (iphone6, ios9.3.2):

- 3s+ in safari, and crash when the root element has more than 500 child nodes
- crash in webview

![timeline](http://beauty.nosdn.127.net/beauty/img/6b32b38d-7f18-43a7-b587-06923db9dedb.png)

**Comment by yyx990803** [comment]
1. There should be a warning telling you not to use `new Vue()` on body.

2. The in-browser compile process can be completely eliminated if you pre-compile the templates using `vue-loader` or `vueify`. If you are really concerned about performance, that's what you should do.

3. Make sure to measure performance with the production build.

4. Need more information. The stack trace is helpful but you are not showing the full trace. Also need the template that you are compiling to see where the problem is.

**Comment by maggiehe** [comment]
Thanks for your reply. 

- We use new Vue( ) on a div which contains most of the information, it’s not a good practice but currently we need Vue directives for lazy-load. And the difference is not big between the dev and production build.
- We refactored the nunjucks template out into multiple Vue components, and resolved the crash problem. But the performance still looks not so good especially in the problematic phone. May be its a problem on a particular version of ios?
  1. safari in iphone6s, ios10.1.1: 360ms -> 324ms
  2. safari in the problematic iphone6, ios9.3.2: 3s -> 1s
- Hope the [chrome timeline data](http://paopao.nosdn.127.net/e34851ec47e57dd7092ae4ebeee0255d.json) can be helpful.

**Comment by HerringtonDarkholme** [comment]
Hi @maggiehe, thanks for reporting this. I have pursued your timeline data. I think there is some irrelevant chrome extension JavaScript, let's just ignore it.

There is two main performance bottlenecks for your application:

1. Vue template compilation. Parsing templates costs about 200ms on Chrome (probably desktop version, I guess). It seems you have quite large HTML corpse. There is no hot function in compilation: `handleStartTag`, `chars` are called repeatedly and execution time naturally piles up.  

This is quite hard to optimize without precompilation. Even if we can squeeze out the last bit of performance from parsing, the performance gain might still fails to match the fattening DOM size . I do recommend you to consider precompilation or selectively compiles part of your static page.

2. Vue instance creation. This is more promising in optimizing. See this graph.

![screen shot 2016-12-07 at 4 59 28 pm](https://cloud.githubusercontent.com/assets/2883231/20961738/4f34d47a-bca1-11e6-8322-3cbf836c9e74.png)

Most time is consumed in `createElement`, which is largely composed of a series of **yellow block** in the graph: `setAttribute` and `updateAttribute` native dom call.

>   we need Vue directives for lazy-load

It might be your directives that slow down the `createElm`.

I don't know your specific situation, but I would recommend extract the directives into vanilla JavaScript module and provide other callbacks to notify Vue instances. 

Hope this will be helpful!

**Comment by maggiehe** [comment]
Thanks for your help, We have resolved our problem with precompilation.

It seems the directives does not affect the performance, there are still frequent setAttr, updateAttr operations with pure static html.

I want to close this issue because I tested this large page with another framework, the two frameworks have the same order of magnitude in build time. Although in this case Vue spent more time, it’s far from enough to be used to measure the performance:)
